File: python/test/01-operations.py
#!/bin/bash

##############################################################################
# BOOTSTRAP
#
# Use python3 or python, whichever is available.
# (See https://unix.stackexchange.com/questions/20880)
#
if "true" : '''\'
then
    exec $(command -v python3 || command -v python) "$0" "$@"
    exit 127
fi
'''


##############################################################################
# PYTHON CODE BEGINS HERE

import sys
import liteexpr

PREOPS = [ "!", "~", "+", "-", "++", "--" ]
POSTOPS = [ "++", "--" ]
BINARYOPS = [
    '**', '*', '/', '%', '+', '-', '<<', '>>', '>>>', '<', '<=', '>', '>=',
    '==', '!=', '&', '^', '|', '&&', '||'
]


def main():
    nan = float("NaN")
    inf = float("Inf")
    ninf = float("-Inf")

    for i in [ 0, 7, 3.5, -7, -3.5, nan, inf, ninf, '""', '"x"' ]:
        for op in PREOPS:
            symbols = liteexpr.SymbolTable({ "i":i })
            expr = f"{op}i"

            try:
                result = liteexpr.eval(expr, symbols)
            except liteexpr.Error as e:
                result = str(e)

            print(f"{'':>4} {op:>3} {i:>4} => {result}")

    for i in [ 0, 7, 3.5, -7, -3.5, nan, inf, ninf, '""', '"x"' ]:
        for op in POSTOPS:
            symbols = liteexpr.SymbolTable({ "i":i })
            expr = f"i{op}"

            try:
                result = liteexpr.eval(expr, symbols)
            except liteexpr.Error as e:
                result = str(e)

            print(f"{i:>4} {op:>3} {'':>4} => {result}")

    for i in [ 0, 0.0, 7, 3.5, -7, -3.5, nan, inf, ninf, '""', '"x"' ]:
        for op in BINARYOPS:
            for j in [ 0, 3, 3.5, -4, -4.5, nan, inf, ninf, '"x"', '"y"' ]:
                symbols = liteexpr.SymbolTable({ "i":i, "j":j })
                expr = f"i {op} j"

                try:
                    result = liteexpr.eval(expr, symbols)
                except liteexpr.Error as e:
                    result = str(e)

                print(f"{i:>4} {op:>3} {j:>4} => {result}")


##############################################################################
# ENTRY POINT

if __name__ == "__main__":
    main()


# vim:filetype=python:

File: cpp/test/01-operations.cpp
#include <vector>
#include <string>
#include <memory>
#include <iostream>
#include "liteexpr.h"

using namespace std;

vector<string> PREOPS = { "!", "~", "+", "-", "++", "--" };
vector<string> POSTOPS = { "++", "--" };
vector<string> BINARYOPS = {
    "**", "*", "/", "%", "+", "-", "<<", ">>", ">>>", "<", "<=", ">", ">=",
    "==", "!=", "&", "^", "|", "&&", "||"
};

const double NaN = 0.0/0.0;
const double Inf = 1.0/0.0;


int main(int argc, const char* argv[]) {
    liteexpr::ARRAY list0 = liteexpr::make_array({
        liteexpr::make_value(0),
        liteexpr::make_value(7),
        liteexpr::make_value(3.5),
        liteexpr::make_value(-7),
        liteexpr::make_value(-3.5),
        liteexpr::make_value(NaN),
        liteexpr::make_value(Inf),
        liteexpr::make_value(-Inf),
        liteexpr::make_value("\"\""),
        liteexpr::make_value("\"x\""),
    });
    liteexpr::ARRAY list1 = liteexpr::make_array({
        liteexpr::make_value(0),
        liteexpr::make_value(0.0),
        liteexpr::make_value(7),
        liteexpr::make_value(3.5),
        liteexpr::make_value(-7),
        liteexpr::make_value(-3.5),
        liteexpr::make_value(NaN),
        liteexpr::make_value(Inf),
        liteexpr::make_value(-Inf),
        liteexpr::make_value("\"\""),
        liteexpr::make_value("\"x\""),
    });
    liteexpr::ARRAY list2 = liteexpr::make_array({
        liteexpr::make_value(0),
        liteexpr::make_value(3),
        liteexpr::make_value(3.5),
        liteexpr::make_value(-4),
        liteexpr::make_value(-4.5),
        liteexpr::make_value(NaN),
        liteexpr::make_value(Inf),
        liteexpr::make_value(-Inf),
        liteexpr::make_value("\"x\""),
        liteexpr::make_value("\"y\""),
    });

    for(liteexpr::VALUE i : list0->native()) {
        for(string op : PREOPS) {
            string expr = op + "i";
            string result;
            liteexpr::SYMBOLS symbols = liteexpr::make_symbols({
                { "i", i },
            });

            try {
                result = liteexpr::eval(expr, symbols)->svalue();
            }
            catch(liteexpr::Error e) {
                result = std::string(e);
            }

            cout << setw(4) << right << "" << " "
                 << setw(3) << right << op << " "
                 << setw(4) << right << i->svalue() << " => "
                 << result << endl;
        }
    }

    for(liteexpr::VALUE i : list0->native()) {
        for(string op : POSTOPS) {
            string expr = string("i") + op;
            string result;
            liteexpr::SYMBOLS symbols = liteexpr::make_symbols({
                { "i", i },
            });

            try {
                result = liteexpr::eval(expr, symbols)->svalue();
            }
            catch(liteexpr::Error e) {
                result = std::string(e);
            }

            cout << setw(4) << right << i->svalue() << " "
                 << setw(3) << right << op << " "
                 << setw(4) << right << "" << " => "
                 << result << endl;
        }
    }

    for(liteexpr::VALUE i : list1->native()) {
        for(string op : BINARYOPS) {
            for(liteexpr::VALUE j : list2->native()) {
                string expr = string("i") + op + "j";
                string result;
                liteexpr::SYMBOLS symbols = liteexpr::make_symbols({
                    { "i", i },
                    { "j", j },
                });

                try {
                    result = liteexpr::eval(expr, symbols)->svalue();
                }
                catch(liteexpr::Error e) {
                    result = std::string(e);
                }

                cout << setw(4) << right << i->svalue() << " "
                     << setw(3) << right << op << " "
                     << setw(4) << right << j->svalue() << " => "
                     << result << endl;
            }
        }
    }

    return 0;
}
